libsolidity/codegen/ir/IRGenerator.cpp

```diff
@@ -238,64 +244,158 @@ string IRGenerator::generateInitialAssignment(VariableDeclaration const& _varDec
+				*(arguments->at(i)),
+				*(baseConstructor->parameters()[i]->type())
+			).commaSeparatedList();
+
+		return params;
+	};
+
+	map<ContractDefinition const*, string> constructorParams;
+	if (FunctionDefinition const* constructor = _contract.constructor())
+		for (auto const& modifier: constructor->modifiers())
+		{
+			FunctionDefinition const* baseConstructor = dynamic_cast<ContractDefinition const*>(
+				modifier->name()->annotation().referencedDeclaration
+			)->constructor();
 
-	FunctionDefinition const* constructor = _contract.constructor();
-	if (constructor && !constructor->isPayable())
-		out << callValueCheck();
+			if (baseConstructor && modifier->arguments() && !modifier->arguments()->empty())
+				constructorParams[dynamic_cast<ContractDefinition const*>(baseConstructor->scope())] =
+					listParamNames(baseConstructor, modifier->arguments());
+		}
 
-	for (ContractDefinition const* contract: reverse(_contract.annotation().linearizedBaseContracts))
+	for (ASTPointer<InheritanceSpecifier> const& base: _contract.baseContracts())
 	{
-		out <<
-			"\n// Begin state variable initialization for contract \"" <<
-			contract->name() <<
-			"\" (" <<
-			contract->stateVariables().size() <<
-			" variables)\n";
-
-		IRGeneratorForStatements generator{m_context, m_utils};
-		for (VariableDeclaration const* variable: contract->stateVariables())
-			if (!variable->isConstant() && !variable->immutable())
-				generator.initializeStateVar(*variable);
-		out << generator.code();
-
-		out << "// End state variable initialization for contract \"" << contract->name() << "\".\n";
+		FunctionDefinition const* baseConstructor = dynamic_cast<ContractDefinition const*>(
+			base->name().annotation().referencedDeclaration
+		)->constructor();
+
+		if (baseConstructor && base->arguments() && !base->arguments()->empty())
+			constructorParams[dynamic_cast<ContractDefinition const*>(baseConstructor->scope())] =
+				listParamNames(baseConstructor, base->arguments());
 	}
 
-	if (constructor)
-	{
-		ABIFunctions abiFunctions(m_evmVersion, m_context.revertStrings(), m_context.functionCollector());
-		unsigned paramVars = make_shared<TupleType>(constructor->functionType(false)->parameterTypes())->sizeOnStack();
+	return {generator.code(), constructorParams};
+}
 
-		Whiskers t(R"X(
-			let programSize := datasize("<object>")
-			let argSize := sub(codesize(), programSize)
+string IRGenerator::initStateVariables(ContractDefinition const& _contract)
+{
+	IRGeneratorForStatements generator{m_context, m_utils};
+	for (VariableDeclaration const* variable: _contract.stateVariables())
+		if (!variable->isConstant() && !variable->immutable())
+			generator.initializeStateVar(*variable);
+
+	return generator.code();
+}
 
-			let memoryDataOffset := <allocate>(argSize)
-			codecopy(memoryDataOffset, programSize, argSize)
+void IRGenerator::generateImplicitConstructors(ContractDefinition const& _contract)
+{
+	auto listAllParams = [&](
```

Comment:
That's the only thing left that I do not yet fully understand :)

Reply:
Ok, Let me try explaining.
This is used in two different situations:
1. List all parameters for a constructor call
 - Here we need to take care of order of parameters, it should list first parameters of user defined constructor, than base arguments in order which map stores them. In this case second parameter is next constructor to be called, so we add parameters of that one first.
2. List base constructor arguments for constructor function definition
 - Here we need only to list base constructor arguments and they are sorted as map stores them. In this case second parameter is nullptr

Hope it makes it clear :)